chainer/testing/function.py

```diff
@@ -0,0 +1,530 @@
+        """Tests forward computation."""
+
+        self.backend_config = backend_config
+        self.before_test('test_forward')
+
+        cpu_inputs = self._generate_inputs()
+        inputs_copied = [_.copy() for _ in cpu_inputs]
+
+        # Compute expected outputs
+        cpu_expected = self._forward_expected(cpu_inputs)
+        inputs = backend_config.get_array(cpu_inputs)
+        inputs = self._to_noncontiguous_as_needed(inputs)
+
+        # Compute actual outputs
+        outputs = self._forward(
+            tuple(chainer.Variable(a) for a in inputs),
+            backend_config)
+
+        # Check inputs has not changed
+        for i in range(len(inputs)):
+            with FunctionTestError.raise_if_fail(
+                    'Input array has been modified during forward.'):
+                array_module.assert_allclose(
+                    inputs_copied[i], inputs[i], atol=0, rtol=0)
+
+        # Check all outputs are Variable instance
+        for out in outputs:
+            FunctionTestError.check(
+                isinstance(out, chainer.Variable),
+                'Output of forward is not a Variable instance.')
+
+        def format_array_props(msg, expected_arrays, arrays):
+            def format_array(arr):
+                return '{}:{}'.format(arr.shape, arr.dtype.name)
+
+            def format_arrays(arrs):
+                return ', '.join([format_array(a) for a in arrs])
+
+            return (
+                '{}\n'
+                'Expected shapes and dtypes: {}\n'
+                'Actual shapes and dtypes:   {}'.format(
+                    msg,
+                    format_arrays(expected_arrays),
+                    format_arrays(arrays)))
+
+        # Check all outputs are equal to expected values
+        FunctionTestError.check(
+            len(outputs) == len(cpu_expected),
+            'Number of forward outputs is incorrect.')
+        dtypes_match = all([
+            y.shape == ye.shape for y, ye in zip(outputs, cpu_expected)])
+        shapes_match = all([
+            y.shape == ye.shape for y, ye in zip(outputs, cpu_expected)])
+        if not (shapes_match and dtypes_match):
+            FunctionTestError.fail(
+                format_array_props(
+                    'Shapes and/or dtypes of forward output is incorrect.',
+                    cpu_expected, outputs))
+        for i in range(len(outputs)):
+            with FunctionTestError.raise_if_fail(
+                    'Forward output does not match the expected value.\n'
+                    'at outputs[{}] (of {})'.format(i, len(outputs))):
+                array_module.assert_allclose(
+                    cpu_expected[i], outputs[i].data,
+                    **self.check_forward_options)
+
+    @_copy_to_test_case(enable_attr_name='backward_test')
+    def test_backward(self, backend_config):
+        """Tests backward computation."""
+
+        # avoid cyclic import
+        from chainer import gradient_check
```

Comment:
Let's add a TODO to not require this workaround.

Reply:
I think this is fine. See `unary_math_function_unittest`.